File: core/src/main/java/io/github/msameer0/rhythmicrush/game/gameplay/interactables/portals/CubePortal.java
package io.github.msameer0.rhythmicrush.game.gameplay.interactables.portals;

import io.github.msameer0.rhythmicrush.game.registries.Registry;

/**
 * A portal that transforms the player into a cube-based movement mode upon contact.
 * <p>
 * When triggered, this portal replaces the current player instance with a cube player
 * retrieved from the world's object pool.
 */
@Registry(id = "cube_portal")
public class CubePortal extends AbstractPortal {

    /**
     * Constructs a new CubePortal at the specified coordinates.
     *
     * @param x the x-coordinate of the portal's position
     * @param y the y-coordinate of the portal's position
     */
    public CubePortal(float x, float y) {
        super(x, y);
        this.type = PortalType.CUBE;
    }

    /**
     * No-arg constructor for pooling — call init() before use.
     */
    public CubePortal() {
        super();
        this.type = PortalType.CUBE;
    }

    @Override
    public CubePortal init(float x, float y) {
        super.init(x, y);
        this.type = PortalType.CUBE;
        return this;
    }
}

File: core/src/main/kotlin/io/github/msameer0/rhythmicrush/game/gameplay/interactables/portals/CubePortal.kt
package io.github.msameer0.rhythmicrush.game.gameplay.interactables.portals

import io.github.msameer0.rhythmicrush.game.registries.Registry

@Registry(id = "cube_portal")
class CubePortal : AbstractPortal {
    constructor(x: Float, y: Float) : super(x, y) {
        this.type = PortalType.CUBE
    }

    constructor() : super() {
        this.type = PortalType.CUBE
    }

    override fun init(x: Float, y: Float, rotation: Float): CubePortal {
        super.init(x, y, rotation)
        this.type = PortalType.CUBE
        return this
    }

    override fun init(x: Float, y: Float): CubePortal {
        return init(x, y, 0f)
    }
}
